### PR TITLE
Add Howto/Introduction document

### PR DESCRIPTION
**Pull request by sigiesec**

### CODE REVIEW

File: Introduction.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+=============================
+
+The idea of the Service IDL is to provide the means to define Service APIs. We use the term "Service API" to denote 
+APIs that are aimed at remote usage, i.e. between different machines or different processes on the same machine. Service 
+APIs are therefore to be distinguished from object-oriented APIs that are intended for local, intra-process usage only.
+However, Service APIs should provide location transparency, i.e. the direct user should not know whether the provider 
+is located in the same process, a different process on the same machine or on a different machine. Service APIs should
+therefore be provided as regular interfaces in the respective programming language. The implementation of such an interface
+may either be a local implementation, or a *proxy* that typically corresponds with a *dispatcher* on the provider side via a remote
+appplication protocol.
```
Comment (huttenlocher) [inline Introduction.md]: Confusing line break after "remote", probably unnecessary.

File: Introduction.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+Currently, the following generation targets are supported:
+* in general, C++, JVM/Java and CLR/.NET are supported as target technologies
+* API generators for each of the target technologies (the C++ generator makes use of the proprietary BTC.CAB.Commons framework)
+* proxy/dispatcher generators exist for the BTC.CAB.ServiceComm service communication APIs in each of the target technologies. 
+  The BTC.CAB.ServiceComm APIs are proprietary as of November 2018, but might be made available as Open Source in the future.
+* the remote protocol is generated as Google Protocol Buffers messages
```
Comment (huttenlocher) [inline Introduction.md]: Capital "The" at the beginning of the new sentence (after period).

File: Introduction.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+Additional generators for the existing or additional target technologies could be added. Currently, there is no extension mechanism
+that allows an external addition, i.e. without modifying the existing code, but this could be easily added using the Equinox/Eclipse
```
Comment (huttenlocher) [inline Introduction.md]: Unnecessary line break after "mechanism".

File: Introduction.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+Currently, the following generation targets are supported:
+* in general, C++, JVM/Java and CLR/.NET are supported as target technologies
+* API generators for each of the target technologies (the C++ generator makes use of the proprietary BTC.CAB.Commons framework)
+* proxy/dispatcher generators exist for the BTC.CAB.ServiceComm service communication APIs in each of the target technologies. 
+  The BTC.CAB.ServiceComm APIs are proprietary as of November 2018, but might be made available as Open Source in the future.
+* the remote protocol is generated as Google Protocol Buffers messages
+
+Additional generators for the existing or additional target technologies could be added. Currently, there is no extension mechanism
+that allows an external addition, i.e. without modifying the existing code, but this could be easily added using the Equinox/Eclipse 
+Extension Point facilities.
```
Comment (huttenlocher) [inline Introduction.md]: Same here... maybe it's just the GitHub diff GUI displaying it that way, I don't know.

File: Introduction.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+Relevant architectural guidelines
+=================================
+
+There are several architectural guidelines that are relevant in the context of the Service IDL. Eit
```
Comment (huttenlocher) [inline Introduction.md]: "Eit" - unclear what is this.

File: Introduction.md [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+Architectural guidelines directly driving the design of the Service IDL:
+* An API must be either service-oriented or object-oriented, but not a mixture of both.
+
+Architectural guidelines relevant for the design of the generators:
+* A service-oriented API should be implemented by an adapter to a corresponding object-oriented API
+* A service implementation must not be bound to a specific hosting environment be separated
```
Comment (huttenlocher) [inline Introduction.md]: Unclear end of the sentence "must not be bound... be separated"; seems like a remnant of something.

File: Introduction.md [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+Modules can be nested hierarchically. One module should be designated as the main module. If no module is explicitly 
+marked as the main module, there are some rules to infer an implicit main module, but this behaviour might be deprecated
+in the future and eventually removed.
+
+The main module may have submodules, which contain definitions, but there may not be any definitions outside the main module.
+
+The name of the main module should match the name of the IDL file. In the example above, the main module is named `foo.bar`, 
+therefore the IDL file should be called `foo.bar.idl`.
+
+A module may contain _interfaces_, _data structures_ and _exceptions_.
```
Comment (huttenlocher) [inline Introduction.md]: Maybe rephrase or state more precisely? Currebtly, by IDL grammar a module may also contain nested modules  (this is already stated above), type aliases (keyword "typedef") and enums.

File: Introduction.md [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+Interfaces can be declared within a module. Interface are declared using the `interface` keyword.
+
+An interface has a name (`Test` in the example above) and must specify a `guid`. An empty interface definition looks like this:
+```
+interface Test[guid=384E277A-C343-4F37-B910-C2CE6B37FC8E] {};
+```
+As an advanced feature, interface can extend other interfaces, but this feature is currently discouraged.
+
+An interface may contain _operations_, _data structures_ and _exceptions_.
```
Comment (huttenlocher) [inline Introduction.md]: Same here (enums, typedef), but maybe they belong in the category "data structures", in this case would be OK.

File: Introduction.md [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+Operations
+----------
+
+Operations can be delcared within an interface as follows:
+```
+    DoSomething(in sequence<string> inputSequence) returns void;
+    query GetSomething() returns sequence<failable uuid>;
+```
+
+Operations can be _commands_ (mutating operations) or _queries_ (non-mutating operations), the latter of which are denoted by the `query` keyword.
```
Comment (huttenlocher) [inline Introduction.md]: Currently, as far as I know, there is no real logic bound to this it the generator, so a may mark it as "query" and do whatever I want within the method. Maybe we should introduce some constraint checking etc. Just a remark.

Comment (huttenlocher) [reply Introduction.md]: In addition to that (even more important, for my taste), operations may be "sync" or asynchronous (by default). As return type, either the bare type or a future is used.

File: Introduction.md [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+
+Every operation has
+* a name (`DoSomething` and `QuerySomething` in the example above),
+* an arbitrary number of parameters
+* a return type
+
```
Comment (huttenlocher) [inline Introduction.md]: Also optionally a "raises" declaration to specify one or more raised exceptions.

File: Introduction.md [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+Data structures can be declared as follows:
+```
+struct KeyValuePair {
+    string key;
+    string value;
+};
+```
+
+The declaration specified the name of the data structure, and contains a declaration of any number of _attributes_.
+Each attribute has a _data type_ (see below) and an _attribute name_.
```
Comment (huttenlocher) [inline Introduction.md]: In addition, an attribute may be marked as "optional".

File: Introduction.md [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+* string
+* float
+* double
+* boolean
+* uuid
+
+The following constructs exist to define more advanced types for use as operation parameter and return types:
+* sequence: A sequence represents a finite number of elements of the specified value type. As an advanced use, the sequence 
+  value type can be marked as `failable`, the details of which are beyond the scope of this introduction. An example of a sequence type is `sequence<int16>`.
+* tuple: A tuple represents a combination of a specified number of elements of different value types, similar to an data
```
Comment (huttenlocher) [inline Introduction.md]: This is part of the grammar, but the usage is currently suppressed by a validator error. Tuples (especially deeply nested tuples) have caused a lot of troubles, it was hard to support them right in the protobuf, since there is no proper basic type and tuples are anonymous (requiring a transparent pseudo wrapper etc.) Also, there are no native tuples in Java, forcing us to do some trickery to map them, etc. We should just **ban** tuples once and for all from the IDL grammar!!! Tuple can be represented by a struct without any problem, but they complicate the generator code in multiple ways, there is no real gain but lots of pain with them.

File: Introduction.md [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+* float
+* double
+* boolean
+* uuid
+
+The following constructs exist to define more advanced types for use as operation parameter and return types:
+* sequence: A sequence represents a finite number of elements of the specified value type. As an advanced use, the sequence 
+  value type can be marked as `failable`, the details of which are beyond the scope of this introduction. An example of a sequence type is `sequence<int16>`.
+* tuple: A tuple represents a combination of a specified number of elements of different value types, similar to an data 
+  structure containing unnamed fields. Data structures with named fields should be preferred. An example of a tuple type is `tuple<string, string>`.
```
Comment (huttenlocher) [inline Introduction.md]: > Data structures with named fields should be preferred.

Yes, yes, YES! No tuples, no tears.

File: Introduction.md [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+```
+
+The first exception (`FooException`) does not declare a supertype, which leaves it to the generator to use an 
+implementation-defined supertype. The second exception (`BarException`) derives from `FooException`, i.e. when 
+catching FooException, instances of BarException will also be caught.
+
+Note that there are curly braces within the exception declaration. The IDL syntax already supports the declaration
+of custom attributes, but this is not currently supported properly by the generators (as of version 1.0.0). You 
+will yield invalid code if an exception declares custom attributes.
+
```
Comment (huttenlocher) [inline Introduction.md]: There is also the possibility to use the "ref" keyword to refer to an exception defined outside... I currently know no real use case for this, it's just in the grammar. I think it was originally intended to introduce a name to resolve this type in Xtext, but to use an user-defined string going into generated code (e.g. to associate a type with some CAB exception).

File: Introduction.md [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+
+Workflows
+=========
+
+After you have written an IDL file, you can use apply the code generator to it. To get an idea what is happening, we
```
Comment (huttenlocher) [inline Introduction.md]: "can use apply"... take any two...

File: Introduction.md [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+
+
+Continous delivery workflow
+---------------------------
+
+On the BTC internal CAB Jenkins (https://ci.bop-dev.de) that provides continuous integration & delivery services, a dedicated support has been
+implemented to simplify building and publishing artifacts from an IDL file as much as possible. Please refer to the documentation of ci.bop-dev.de 
+for general information on using the Jenkins, and on how the release process works.
+
+An example can be found under https://bitbucket.e-konzern.de/projects/BTCCABCOM/repos/serviceidl-demo/browse
```
Comment (huttenlocher) [inline Introduction.md]: Make hyperlink?

File: Introduction.md [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+java -jar com.btc.serviceidl.plainjava-1.0.0.jar input.idl -dotnetOutputPath outDotNet
+```
+
+Currently, the generator generates all artifacts (API, Proxy, Dispatcher, ...) for all target technologies (C++, C#/.NET, Java) by default.
+
+
+Continous delivery workflow
+---------------------------
+
+On the BTC internal CAB Jenkins (https://ci.bop-dev.de) that provides continuous integration & delivery services, a dedicated support has been
```
Comment (huttenlocher) [inline Introduction.md]: Make hyperlink?

File: Introduction.md [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+    // java_idl_build already publishes artifacts
+}
+```
+
+This example contains section for C++, .NET and Java, which use the IDL generator to generate code in each of the technologies, 
+build the generated code, and publish technology-specific packages. The Jenkinsfile may contain any subset of these sections, 
+if you only need packages for some of these technologies.
+
+Optionally, you can also add a `.generator` file that contains non-default settings for the IDL generators, e.g. version overrides
+for the target technologies.
```
Comment (huttenlocher) [inline Introduction.md]: Unnecessary line break?

File: Introduction.md (hunk 2 above)
Comment (sigiesec) [reply Introduction.md]: Fixed

File: Introduction.md (hunk 5 above)
Comment (sigiesec) [reply Introduction.md]: Hm, something that should have started with "Either", but I don't remember more ;) Removed this for now.

File: Introduction.md (hunk 12 above)
Comment (sigiesec) [reply Introduction.md]: Right, I forgot about this. Removed references to tuples.

File: Introduction.md (hunk 16 above)
Comment (sigiesec) [reply Introduction.md]: URLs are automatically rendered as links with the URL as text. The (...)[...] syntax can be used to specify a different text to display, but I don't think this makes sense in this case.

File: Introduction.md (hunk 17 above)
Comment (sigiesec) [reply Introduction.md]: Changed to use "BTC internal CAB Jenkins" as the link text.

File: Introduction.md (hunk 14 above)
Comment (sigiesec) [reply Introduction.md]: I left this out deliberately.  How would you handle the different target technologies? I think this cannot be implemented, and I think it might be removed from the grammar completely in the future.

File: Introduction.md (hunk 6 above)
Comment (sigiesec) [reply Introduction.md]: Yes, this was left over from a previous version, I removed the last two words

File: Introduction.md (hunk 15 above)
Comment (sigiesec) [reply Introduction.md]: Fixed

File: Introduction.md (hunk 7 above)
Comment (sigiesec) [reply Introduction.md]: Rephrased.

File: Introduction.md (hunk 8 above)
Comment (sigiesec) [reply Introduction.md]: Changed to "data types"

File: Introduction.md (hunk 11 above)
Comment (sigiesec) [reply Introduction.md]: Added this.

File: Introduction.md (hunk 9 above)
Comment (sigiesec) [reply Introduction.md]: Aren't queries at least generated as const in C++?

Comment (sigiesec) [reply Introduction.md]: Added a sentence on synchronous/asynchronous

File: Introduction.md (hunk 10 above)
Comment (sigiesec) [reply Introduction.md]: Added
